internal/merkle/inmemory/tree.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2022 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package inmemory provides an in-memory Merkle tree implementation.
+package inmemory
+
+import (
+	"fmt"
+
+	"github.com/transparency-dev/merkle"
+	"github.com/transparency-dev/merkle/compact"
+	"github.com/transparency-dev/merkle/proof"
+)
+
+// Tree implements an append-only Merkle tree. For testing.
```

Comment:
Can do, but I tend to think that everything is not thread-safe in Go by default. Sort of fits with the concurrency best practices: https://go.dev/doc/effective_go#sharing.

> Do not communicate by sharing memory; instead, share memory by communicating.

So, shared thread-safe objects seem not "standard".

Reply:
Very unrelatedly to this PR (but thanks to this thread for triggering thoughts), I think it's possible to implement an efficient immutable version of this `Tree` type! I.e. instead of having `func (t *Tree) Append(hash byte[])`, we could have `func (t Tree) Append(hash []byte) Tree`. The implementation of `Append` decomposes into a few slice `append` calls. The practical use of it, for example, would be the ability to rollback to the previous state easily.

Reply:
Implemented in #2701. It's cool!